### PR TITLE
fix: add thread error handling to the BSP

### PR DESCRIPTION
**Pull request by robertlaurin**
This adds graceful handling for when creating a new thread raises.  We observed this scenario with an application that was improperly terminating sidekiq which resulted in a `ThreadError`.

mperham/sidekiq#4029
mperham/sidekiq#4738

### CODE REVIEW

File: sdk/lib/opentelemetry/sdk/trace/export/batch_span_processor.rb [hunk 1]
```diff
@@ -177,6 +177,8 @@ def reset_on_fork(restart_thread: true)
             @pid = pid
             spans.clear
             @thread = restart_thread ? Thread.new { work } : nil
+          rescue ThreadError
+            @metrics_reporter.add_to_counter('otel.bsp.failure', labels: { 'reason' => 'ThreadError' })
```
Comment (fbogsany) [inline sdk/lib/opentelemetry/sdk/trace/export/batch_span_processor.rb:181]: Do we also want to log here?

Comment (robertlaurin) [reply sdk/lib/opentelemetry/sdk/trace/export/batch_span_processor.rb:181]: I think so, did you have any thoughts on the counter name?  Is `otel.bsp.failure` fine?

Comment (fbogsany) [reply sdk/lib/opentelemetry/sdk/trace/export/batch_span_processor.rb:181]: `otel.bsp.error` might be better

File: sdk/lib/opentelemetry/sdk/trace/export/batch_span_processor.rb [hunk 2]
```diff
@@ -177,6 +177,9 @@ def reset_on_fork(restart_thread: true)
             @pid = pid
             spans.clear
             @thread = restart_thread ? Thread.new { work } : nil
+          rescue ThreadError => e
+            @metrics_reporter.add_to_counter('otel.bsp.error', labels: { 'reason' => 'ThreadError' })
+            OpenTelemetry.logger.error "OpenTelemetry error in BatchSpanProcessor#reset_on_fork #{e.message}"
```
Comment (mwear) [inline sdk/lib/opentelemetry/sdk/trace/export/batch_span_processor.rb:182]: Are we supposed use `OpenTelemetry.logger.error` or `OpenTelemetry.handle_error` for this purpose (I'm unsure).

Comment (fbogsany) [reply sdk/lib/opentelemetry/sdk/trace/export/batch_span_processor.rb:182]: Great question - I can't think of a good reason *not* to use `OpenTelemetry.handle_error`.

Comment (robertlaurin) [reply sdk/lib/opentelemetry/sdk/trace/export/batch_span_processor.rb]: Updated, I think it makes sense to let the users decide on how to handle this error.  And it has a sane default of logging it.

**Review by fbogsany — APPROVED** [review]

**Pull request merged**